protocol/nats_jetstream/v3/receiver.go

```diff
@@ -0,0 +1,231 @@
+	conn, err := nats.Connect(url, natsOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := NewConsumerFromConn(ctx, conn, jsOpts, opts...)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	c.connOwned = true
+
+	return c, err
+}
+
+// NewConsumerFromConn consumes from filterSubjects given in ConsumerConfig or OrderedConsumerConfig
+// embedded in ConsumerOption. See: WithConsumerConfig(...) and WithOrderedConsumerConfig(...)
+func NewConsumerFromConn(ctx context.Context, conn *nats.Conn, jsOpts []jetstream.JetStreamOpt, opts ...ConsumerOption) (*Consumer, error) {
+	var js jetstream.JetStream
+	var err error
+	if js, err = jetstream.New(conn, jsOpts...); err != nil {
+		return nil, err
+	}
+
+	c := &Consumer{
+		Receiver:      *NewReceiver(),
+		Conn:          conn,
+		JetStream:     js,
+		internalClose: make(chan struct{}, 1),
+	}
+
+	if err = c.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (c *Consumer) OpenInbound(ctx context.Context) error {
+	c.subMtx.Lock()
+	defer c.subMtx.Unlock()
+
+	var consumeContext jetstream.ConsumeContext
+	var err error
+	if err = c.createConsumer(ctx); err != nil {
+		return err
+	}
+	if consumeContext, err = c.JetstreamConsumer.Consume(c.MsgHandler, c.PullConsumeOpt...); err != nil {
+		return err
+	}
+
+	// Wait until external or internal context done
+	select {
+	case <-ctx.Done():
+	case <-c.internalClose:
+	}
+
+	// Finish to consume messages in the queue and close the subscription
+	if consumeContext != nil {
+		consumeContext.Drain()
+	}
+	return nil
+}
+
+// Close implements Closer.Close
+func (c *Consumer) Close(ctx context.Context) error {
+	// Before closing, let's be sure OpenInbound completes
+	// We send a signal to close and then we lock on subMtx in order
+	// to wait OpenInbound to finish draining the queue
+	c.internalClose <- struct{}{}
+	c.subMtx.Lock()
+	defer c.subMtx.Unlock()
+
+	if c.connOwned {
+		c.Conn.Close()
+	}
+
+	close(c.internalClose)
+
+	return nil
+}
+
+func (c *Consumer) applyOptions(opts ...ConsumerOption) error {
+	for _, fn := range opts {
+		if err := fn(c); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// createConsumer creates a consumer based on the configured consumer config
+func (c *Consumer) createConsumer(ctx context.Context) error {
+	var err error
+	var stream string
+	if stream, err = c.getStreamFromSubjects(ctx); err != nil {
+		return err
+	}
+	consumerType := c.getConsumerType()
+	switch consumerType {
+	case ConsumerType_Unknown:
+		return ErrNoConsumerConfig
+	case ConsumerType_Ordinary:
+		c.JetstreamConsumer, err = c.JetStream.CreateOrUpdateConsumer(ctx, stream, *c.ConsumerConfig)
+	case ConsumerType_Ordered:
+		c.JetstreamConsumer, err = c.JetStream.OrderedConsumer(ctx, stream, *c.OrderedConsumerConfig)
+	}
```

Comment:
nit: feels overengineered to introduce types here. The simple "getConsumerType" logic can be copied and used here without introducing new types/const